### ISSUE TITLE
Concourse build output logging options

### ISSUE THREAD

**Issue by billimek** [concourse/concourse#645]
As a log inspector,
I want to have a mechanism to capture concourse job build output to some sort of logging facility (e.g. syslog sink)
So that I can continue to 'log everything' and not miss-out on the build output

**Comment by jtarchie** [comment]
This would be interesting, because the log output is structure based on `get`, `put`, `task`, etc. How would you distinguish that in the output that is forwarded?

**Comment by billimek** [comment]
I don't think it would be necessary to distinguish.  At least for an MVP just getting all the output captured would be a win.

Otherwise, perhaps prefixing the output (maybe with `[get]`, `[put]`, `[task]`, etc) would get us some level of delineation but honestly that seems a bit heavy-handed IMO

**Comment by jtarchie** [comment]
Well another concern might the security concern. With teams, pipelines, and jobs being possibly publicly disabled/enabled, how do you make sure the right info is passed along to the syslogs. You have someone now able to see all the things.

**Comment by vito** [comment]
Depends on where it's configured. Pipeline or team could be a reasonable place. (Depending on if you'd want syslog for one-off builds too.)

**Comment by poida** [comment]
After demonstrating the pipelines I'd been building to the team one, I was asked whether it's possible to add timestamps to the log output like jenkins can (https://wiki.jenkins-ci.org/display/JENKINS/Timestamper). We'd like the ability to add this as a logging option across all tasks/resources in a pipeline, without having to modify the tasks/resources themselves to print out the timestamp.

If this could be implemented by some sort of generic filtering mechanism that gets and returns log output (not sure what technology, HTTP get, websocket, unix pipe, to a docker?), then it could be used to meet a number of different use cases:

- Ship the logs off to another system while returning logs unscathed (identity filter with side effects);
- Add time stamps;
- Remove secrets or other sensitive data that makes it into the logs.

**Comment by petemounce** [comment]
I would love to have the ability to query (structured) build logs centrally. I would use that to
* discover and count common error messages, to detect and prioritise fixes for flaky tests
* profile how long different jobs take, over time
* notify different people/systems when a job completes with success/error

Logs become _much_ more useful when instead of being unstructured strings, they're machine-parseable. JSON is a decent text format, protobuf is a decent binary format. I'd want to see these fields:
* timestamp (unix UTC, _not_ serialised to a human-readable string since that's a rendering concern, not a publishing one)
* pipeline
* plan
* group
* task
* version
* log-line from build - with the ability that, if the build emits a line of structured data, there's the opportunity for a parser to be given to concourse to munge the message here into a sub-object
* ... other useful things

I'm most familiar with the [ELK stack](https://www.elastic.co/webinars/introduction-elk-stack) for doing these things, with http://www.github.com/Yelp/ElastAlert as an automated log-watcher. Docker has [a good hook-up for pushing logs into it via its log-drivers](https://docs.docker.com/engine/admin/logging/overview/).

**Comment by jchesterpivotal** [comment]
I wished for this, or something like this, today.

One of those "why isn't this more like Cloud Foundry?" moments :|

**Comment by marcbachmann** [comment]
Yes logs of builds would be nice.

I have some concept in my head and also did some prototypes to build a plugin system on the frontend. Frontend plugins could listen to specific streams to process the events (client-side reducers that generate the html).

To build that it would need build start & stop events for a specific and the specific build stream. If you don't plan to expose them over syslog, could I use the api to fetch them? How stable do you consider the api? Is there some documentation?

**Comment by spmsaj** [comment]
Did this get prioritized? I couldnt see the link above

**Comment by DenisIzmaylov** [comment]
Would be nice to have it like Build Bot where logs looks pretty same like a real terminal:

![image](https://user-images.githubusercontent.com/1690596/36803646-31f53024-1cc9-11e8-8f0c-03a9d6f68de0.png)

**Comment by steelecliftonberry** [comment]
I'm also curious if this got prioritised. We have a pretty strong need for this. We have centralised logging capturing everything else in our Kubernetes clusters, but Concourse pipeline logs is a little dark spot.

**Comment by jama-pivotal** [comment]
This is currently sitting in our Operations Backlog. Unfortunately, with the current team size and the priorities set for 2018, we don't have anyone dedicated to looking at the Operations backlog of work. This is definitely one of those issues where PRs will be gladly welcomed and reviewed

**Comment by spmsaj** [comment]
How can we help get this prioritized. It does look like quite a lot of customers are wishing for this. So should have been on a priority list. A MVP can be a first phase.

**Comment by aeckard87** [comment]
I'm also putting in a strong need for this

**Comment by Onke** [comment]
Same here, we definitely need a way to save/ship the log files

**Comment by jimconner** [comment]
+1 for this one... I found this issue because I was looking for a way to ship the build logs off to a syslog endpoint - This feature is desirable to us too.

**Comment by lochlainnw** [comment]
+1 on this also. Logs from builds would be a huge help.

**Comment by samgurtman** [comment]
+1 for this also. Our dev teams are requesting this from us and the workarounds are not really feasible.

**Comment by StevenArmstrong** [comment]
+1 build log shipping would be great.

**Comment by petemounce** [comment]
It is. I built it for our CI, and it allows us to quantify flakes and find introduction-points much easier.

**Comment by spmsaj** [comment]
Saw the Metrics Tools - good one. But what I was referring to was something much simpler - to be able to access the concourse runtime logs for a given pipeline within the pipeline itself. Right now, the only way to do it is  to run the pipeline, run fly watch and then pipe that to another process for log handling. As a consumer of concourse, I would like to get the output of my pipeline, not just on the GUI, but tangible to be exported as part of pipeline run and sent to my custom log handler

**Comment by jchesterpivotal** [comment]
It occurred to this morning that a placeholder option would be a "Concourse Build Resource". Both the web UI and `fly` use an API to fetch information about the structure and output events of a build. I assume that a resource could `check` the build list and `get` could use the same API as `fly` and the web UI to fetch.

With some additional transformations and sugar (eg, output both plaintext and json formats) it would be pretty useful for enabling logs to be shipped to other systems. (`get` a build, `put` to a bucket, Splunk, whatever).

**Comment by jchesterpivotal** [comment]
As mused above, [I wrote a Concourse resource that `get`s build information](https://github.com/jchesterpivotal/concourse-build-resource) that you could then `put` somewhere else.

Caveats: It's got at least one known bug, it doesn't handle auth and it's untested (because it's a quick hack to prove a point), but I think it would ease some of the pains in this thread.

**Comment by pivotal-saman-alvi** [comment]
Check #2104 for implementation details.

**Comment by ranjithfd** [comment]
> As a log inspector,
> I want to have a mechanism to capture concourse job build output to some sort of logging facility (e.g. syslog sink)
> So that I can continue to 'log everything' and not miss-out on the build output

Have you achieved this scenario, i would like to set up the sameway. i want collect all the concourse log to splunk . please let me know if you have any steps to take it forward.
